src/cc/bcc_zip.c

```diff
@@ -0,0 +1,357 @@
+  uint32_t magic;
+  // Minimum zip version needed to extract the file.
+  uint16_t min_version;
+  uint16_t flags;
+  uint16_t compression;
+  uint16_t last_modified_time;
+  uint16_t last_modified_date;
+  uint32_t crc;
+  uint32_t compressed_size;
+  uint32_t uncompressed_size;
+  uint16_t file_name_length;
+  uint16_t extra_field_length;
+};
+
+struct bcc_zip_archive {
+  void* data;
+  uint32_t size;
+  uint32_t cd_offset;
+  uint32_t cd_records;
+};
+
+static int check_access(struct bcc_zip_archive* archive, uint32_t offset,
+                        uint32_t size) {
+  if (offset + size > archive->size || offset > offset + size) {
+    return -1;
+  }
+  return 0;
+}
+
+static int read_cd_file_header(struct bcc_zip_archive* archive,
+                               uint32_t* offset,
+                               struct central_directory_file_header* header) {
+  if (check_access(archive, *offset, CD_FILE_HEADER_SIZE)) {
+    return -1;
+  };
+
+  memcpy(header, archive->data + *offset, CD_FILE_HEADER_SIZE - 8);
+  // skip padding
+  memcpy(&header->external_attributes,
+         archive->data + *offset + CD_FILE_HEADER_SIZE - 8, 8);
+
+  *offset += CD_FILE_HEADER_SIZE;
+  return 0;
+}
+
+static int read_local_file_header(struct bcc_zip_archive* archive,
+                                  uint32_t* offset,
+                                  struct local_file_header* header) {
+  if (check_access(archive, *offset, LOCAL_FILE_HEADER_SIZE)) {
+    return -1;
+  };
+
+  memcpy(header, archive->data + *offset, 14);
+  // skip padding
+  memcpy(&header->crc, archive->data + *offset + 14, 16);
+
+  *offset += LOCAL_FILE_HEADER_SIZE;
+  return 0;
+}
+
+// Returns 0 on success, -1 on error and -2 if the eocd indicates
+// the archive uses features which are not supported.
+static int parse_end_of_central_directory(struct bcc_zip_archive* archive,
+                                          uint32_t offset) {
+  struct end_of_central_directory_record eocd;
+  memcpy(&eocd, archive->data + offset, END_OF_CD_RECORD_SIZE);
+
+  if (eocd.magic != END_OF_CD_RECORD_MAGIC) {
+    return -1;
+  }
+
+  if (offset + END_OF_CD_RECORD_SIZE + eocd.comment_length != archive->size) {
+    return -1;
+  }
+
+  if (eocd.this_disk != 0 || eocd.cd_disk != 0 ||
+      eocd.cd_records != eocd.cd_records_total) {
+    // This is a valid eocd, but we only support single-file non-ZIP64 archives.
+    return -2;
+  }
+
+  if (check_access(archive, eocd.cd_offset, eocd.cd_size)) {
+    return -1;
+  }
+
+  archive->cd_offset = eocd.cd_offset;
+  archive->cd_records = eocd.cd_records;
+  return 0;
+}
+
+static int find_central_directory(struct bcc_zip_archive* archive) {
+  if (archive->size <= END_OF_CD_RECORD_SIZE) {
+    return -1;
+  }
+
+  int rc = -1;
+  uint32_t offset = archive->size - END_OF_CD_RECORD_SIZE;
+  int64_t limit = (int64_t)offset - (1 << 16);
+  for (; offset >= 0 && offset > limit && rc == -1; offset--) {
```

Comment:
To confirm my understanding: this loop is necessary because the variable-length "comment" field at end of "end of central directory record" makes it impossible to find the beginning of the record w/o searching?

If so, can you add a comment explaining that `limit = ... (1 << 16)` above is the earliest place "end of central directory record"  can be?

Also, assuming my earlier "check the magic number before memcpying" comment on `parse_end_of_central_directory` makes sense, it should be possible to avoid checking every byte offset since `END_OF_CD_RECORD_MAGIC` is 4 bytes

Reply:
This is correct.

Can you elaborate on not checking every byte offset ? Do you want me to check the magic number and only if it matches call parse_end_of_central_directory ?

Reply:
>   Can you elaborate on not checking every byte offset ? Do you want me to check the magic number and only if it matches call parse_end_of_central_directory ?

By this I meant that you could use something like [Boyer-Moore algorithm's "bad character" rule](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm) to skip over potential starting offsets.  TL;DR of the heuristic:

Assume we're looking for string `1234` in a string of digits (unknowns represented by `X`). First we'll compare to the end of the haystack string:

```
XXXXX1234234
        1234
```

`4234` != `1234`, but instead of comparing against `3423` next, since we know that the string we're searching for ends in `4`, and there was a `4` in wrong place in our initial comparison, we can jump to comparing against whatever index results in that `4` being end of the string. 

But I wasn't expecting you to write any of that yourself, but maybe consider using something like `memmem` helper. Regardless, I don't feel strongly about doing any change here until since there's no data showing that this is slow. Just looping back.